kong/plugins/ldap-auth/asn1.lua

```diff
@@ -1,408 +1,312 @@
+    start = start or 0
+    stop = stop or #der
+    if stop < start or stop > #der then
+      return nil, "invalid offset"
     end
 
-    return pos, elen
-  end,
-
-  decodeSeq = function(self, encStr, len, pos)
-    local seq = {}
-    local sPos = 1
-    local sStr
-
-    pos, sStr = bunpack(encStr, "A" .. len, pos)
-
-    while (sPos < len) do
-      local newSeq
+    local s_der = ffi_cast("const unsigned char *", der)
+    strpp[0] = s_der + start
 
-      sPos, newSeq = self:decode(sStr, sPos)
-      if not newSeq and self.stoponerror then
-        break
-      end
-
-      insert(seq, newSeq)
-    end
-
-    return pos, seq
-  end,
-
-  decode_oid_component = function(encStr, pos)
-    local octet
-    local n = 0
-
-    repeat
-      pos, octet = bunpack(encStr, "b", pos)
-      n = n * 128 + bit.band(0x7F, octet)
-    until octet < 128
-
-    return pos, n
-  end,
-
-  decodeOID = function(self, encStr, len, pos)
-    local last
-    local oid = {}
-    local octet
-
-    last = pos + len - 1
-    if pos <= last then
-      oid._snmp = "06"
-      pos, octet = bunpack(encStr, "C", pos)
-      oid[2] = math.fmod(octet, 40)
-      octet = octet - oid[2]
-      oid[1] = octet/40
+    local ret = C.ASN1_get_object(strpp, lenp, tagp, classp, stop - start + 1)
+    if band(ret, 0x80) == 0x80 then
+      return nil, "der with error encoding: " .. ret
     end
 
-    while pos <= last do
-      local c
-      pos, c = self.decode_oid_component(encStr, pos)
-      oid[#oid + 1] = c
+    local cons = false
+    if band(ret, 0x20) == 0x20 then
+      cons = true
     end
 
-    return pos, oid
-  end,
+    local obj = {
+      tag = tagp[0],
+      class = classp[0],
+      len = tonumber(lenp[0]),
+      offset = strpp[0] - s_der,
+      hl = strpp[0] - s_der - start,
+      cons = cons,
+    }
 
-  decodeInt = function(encStr, len, pos)
-    local hexStr
-
-    pos, hexStr = bunpack(encStr, "X" .. len, pos)
-
-    local value = tonumber(hexStr, 16)
-    if value >= math.pow(256, len)/2 then
-      value = value - math.pow(256, len)
-    end
-
-    return pos, value
+    return obj
   end
-}
-
-_M.ASN1Encoder = {
-  new = function(self)
-    local o = {}
-    setmetatable(o, self)
-    self.__index = self
-    o:registerBaseEncoders()
-    return o
-  end,
-
-  encodeSeq = function(self, seqData)
-    return bpack("XAA" , "30", self.encodeLength(#seqData), seqData)
-  end,
-
-  encode = function(self, val)
-    local vtype = type(val)
-
-    if self.encoder[vtype] then
-      return self.encoder[vtype](self,val)
-    end
-  end,
-
-  registerTagEncoders = function(self, tagEncoders)
-    self:registerBaseEncoders()
+end
+_M.get_object = asn1_get_object
 
-    for k, v in pairs(tagEncoders) do
-      self.encoder[k] = v
-    end
-  end,
 
-  registerBaseEncoders = function(self)
-    self.encoder = {}
+local function asn1_put_object(tag, class, constructed, data, len)
+  len = type(data) == "string" and #data or len or 0
+  if len < 0 then
```

Comment:
```suggestion
  if len <= 0 then
```

Reply:
The `length` is ignored if `constructed` = `2`
https://github.com/openssl/openssl/blob/e915c3f5381cd38ebdc1824c3ba9896ea7160103/crypto/asn1/asn1_lib.c#L183-L186
```suggestion
  if constructed ~= 2 and len <= 0 then
```

Reply:
Thanks, it makes sense! The `len` could not be -1 in `ffi.new("unsigned char[?]", len)`